### PR TITLE
Refactor image search functions to use Get and TryGet terminology

### PR DESCRIPTION
**Pull request by PunkPun**
fixes

Screenshot 2022-09-12 at 11 32 51

### CODE REVIEW

File: OpenRA.Game/Graphics/ChromeProvider.cs [hunk 1]
```diff
@@ -154,12 +154,17 @@ public static Sprite GetImage(string collectionName, string imageName)
...
 			{
-				Log.Write("debug", "Could not find collection '{0}'", collectionName);
+				Log.Write("debug", $"Could not find collection '{collectionName}'");
 				return null;
 			}
 
 			if (!collection.Regions.TryGetValue(imageName, out var mi))
+			{
+				if (logNotFound)
+					Log.Write("debug", $"Could not find image '{imageName}' in collection '{collectionName}'");
```
Comment (dragunoff) [inline OpenRA.Game/Graphics/ChromeProvider.cs]: This results in several new lines in the log:
```
Could not find image 'corner-tl' in collection 'scrollitem'
Could not find image 'border-t' in collection 'scrollitem'
Could not find image 'corner-tr' in collection 'scrollitem'
Could not find image 'border-l' in collection 'scrollitem'
Could not find image 'background' in collection 'scrollitem'
Could not find image 'border-r' in collection 'scrollitem'
Could not find image 'corner-bl' in collection 'scrollitem'
Could not find image 'border-b' in collection 'scrollitem'
Could not find image 'corner-br' in collection 'scrollitem'
Could not find image 'corner-tl' in collection 'scrollitem'
Could not find image 'border-t' in collection 'scrollitem'
Could not find image 'corner-tr' in collection 'scrollitem'
Could not find image 'border-l' in collection 'scrollitem'
Could not find image 'background' in collection 'scrollitem'
Could not find image 'border-r' in collection 'scrollitem'
Could not find image 'corner-bl' in collection 'scrollitem'
Could not find image 'border-b' in collection 'scrollitem'
Could not find image 'corner-br' in collection 'scrollitem'
```

**Review by abcdefg30 — COMMENTED** [review]
This seems weird. Why are we searching for non-existing images in the first place?

**Comment by PunkPun** [comment]
> This seems weird. Why are we searching for non-existing images in the first place?

Because we want UI to support all state images without having a ton of redundant yaml

**Comment by dragunoff** [comment]
> This seems weird. Why are we searching for non-existing images in the first place?

Because we want to avoid declaring chrome images for every state of an UI element (disabled, focused, pressed, hover) when we don't need them and fall back to the base image :shrug: The current plumbing seems not great at that and this PR is trying to work -around that but I feel like a bool flag is not the right call here.

File: OpenRA.Game/Graphics/ChromeProvider.cs (hunk 1 above)
Comment (PunkPun) [reply OpenRA.Game/Graphics/ChromeProvider.cs]: This log should have been here from the very start. We're going to have to investigate this case by case

**Comment by PunkPun** [comment]
> The current plumbing seems not great at that and this PR is trying to work -around that but I feel like a bool flag is not the right call here.

I think bool flag is the only decent way to avoid code duplication. But perhaps we should move `GetCachedStatefulImage` and related functions to `ChromeProvider.cs` next to the `GetImage` funtion?

File: OpenRA.Game/Graphics/ChromeProvider.cs (hunk 1 above)
Comment (PunkPun) [reply OpenRA.Game/Graphics/ChromeProvider.cs]: fixed

**Comment by abcdefg30** [comment]
> Because we want to avoid declaring chrome images for every state of an UI element (disabled, focused, pressed, hover) when we don't need them and fall back to the base image 🤷 The current plumbing seems not great at that and this PR is trying to work -around that but I feel like a bool flag is not the right call here.

That sounds like the code should not be asking for every state then, doesn't it? 🤔

**Comment by PunkPun** [comment]
> That sounds like the code should not be asking for every state then, doesn't it? 🤔

it's not, see `GetImage` and `GetPanelImages` functions. Both of them are altered in this pr

**Review by RoosterDragon — DISMISSED** [review]
The flag makes sense to me, the goal of the lookups is to use a specific image if one is defined, but otherwise use the base image.

In this scenario if the specific image isn't defined we're expecting the base image to be used. So logging a message if the specific image is missing isn't helpful.

**Comment by PunkPun** [comment]
rebased

**Comment by pchote** [comment]
IMO exposing `TryGetImage` methods would present a cleaner API than the `logNotFound` bool.

**Comment by PunkPun** [comment]
We allow the program to function even if we didn't find the image, this is just for logging. If anything using TryGet would make it more confusing. However I'm not saying that the current way the system functions is correct / desired

**Comment by abcdefg30** [comment]
The problem with `TryGet` is that it moves logging the error to the caller, which we don't want in all cases. A compromise would be adding `TryGet*` methods but doing that while keeping the old `Get*` methods:
- Cases where we don't want to log an error can directly call the `TryGet` method
- Cases where we want a log to be written call the old `Get` method (so most of the callers see no change)
- Internally, the `Get` method utilizes `TryGet` and logs the error if `TryGet` returns false

Where the error the `Get` method handles is the one changed by the parameter in this PR. All other errors remain as-is.

**Comment by PunkPun** [comment]
separating into 2 methods requires duplication of code, or for us to keep the parameter which would defeat the purpose of having another function

**Comment by PunkPun** [comment]
But I could behind renaming `GetImage` to `TryGetImage` and `GetPanelImages` to `TryGetPanelImages`, because that's what they essentially do

**Comment by PunkPun** [comment]
Actually going through all the usecases of GetImage I see that we could just remove logging and crash instead. The code crashes regardless further on as null checks don't exist

**Comment by PunkPun** [comment]
I've refactored it to use the `Get` `TryGet` terminology, as well as made it crash on `Get` instead of just logging

**Comment by PunkPun** [comment]
rebased

**Comment by Mailaender** [comment]
> Loading mod: modcontent
Exception of type `System.ArgumentException`: Sprite `panel-rule/corner-tl` was not found.
   at OpenRA.Graphics.ChromeProvider.GetImage(String collectionName, String imageName) in /home/matthias/Entwicklung/OpenRA/OpenRA/OpenRA.Game/Graphics/ChromeProvider.cs:line 150
   at OpenRA.Graphics.ChromeProvider.TryGetPanelImages(String collectionName) in /home/matthias/Entwicklung/OpenRA/OpenRA/OpenRA.Game/Graphics/ChromeProvider.cs:line 195
   at OpenRA.Mods.Common.Widgets.BackgroundWidget.Draw() in /home/matthias/Entwicklung/OpenRA/OpenRA/OpenRA.Mods.Common/Widgets/BackgroundWidget.cs:line 26
   at OpenRA.Widgets.Widget.DrawOuter() in /home/matthias/Entwicklung/OpenRA/OpenRA/OpenRA.Game/Widgets/Widget.cs:line 479
   at OpenRA.Widgets.Widget.DrawOuter() in /home/matthias/Entwicklung/OpenRA/OpenRA/OpenRA.Game/Widgets/Widget.cs:line 481
   at OpenRA.Widgets.Widget.DrawOuter() in /home/matthias/Entwicklung/OpenRA/OpenRA/OpenRA.Game/Widgets/Widget.cs:line 481
   at OpenRA.Game.RenderTick() in /home/matthias/Entwicklung/OpenRA/OpenRA/OpenRA.Game/Game.cs:line 733
   at OpenRA.Game.Loop() in /home/matthias/Entwicklung/OpenRA/OpenRA/OpenRA.Game/Game.cs:line 841
   at OpenRA.Game.Run() in /home/matthias/Entwicklung/OpenRA/OpenRA/OpenRA.Game/Game.cs:line 873
   at OpenRA.Game.InitializeAndRun(String[] args) in /home/matthias/Entwicklung/OpenRA/OpenRA/OpenRA.Game/Game.cs:line 308
   at OpenRA.Launcher.Program.Main(String[] args) in /home/matthias/Entwicklung/OpenRA/OpenRA/OpenRA.Launcher/Program.cs:line 32

**Comment by PunkPun** [comment]
It appears it found some redundant definitions

**Comment by PunkPun** [comment]
`panel-rule` seems to be as old as time itself (over 9 y. o.) and survived many ui related refactors, it either didn't work from the start or lost it's functionality somewhere years ago

**Review by Mailaender — DISMISSED** [review]

**Comment by PunkPun** [comment]
My bad. After a long investigation into panel rendering code I've found out that `panel-rule` is written correctly and should in theory be functional. Though I've still yet to find the actual visual difference it makes by existing.

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/d0ceb9840e4859fbe6130224fee733b8bf1ef677...7b4fa232fdfda6fdf3b3065773e2f6ee982f8fec)

**Pull request merged**